src/strategy/filesystem.rs

```diff
@@ -0,0 +1,180 @@
+
+impl StrategyFilesystem {
+    /// Strategy label/name.
+    pub const LABEL: &'static str = "filesystem";
+    /// Local filesystem path to finalization marker file.
+    pub const FINALIZATION_MARKER_FILE_PATH: &'static str = "/var/lib/zincati/allowfinalize.json";
+
+    /// Check if finalization is allowed.
+    pub(crate) fn can_finalize(&self) -> Pin<Box<dyn Future<Output = Result<bool, Error>>>> {
+        Box::pin(Self::filesystem_allow_finalization(
+            Self::FINALIZATION_MARKER_FILE_PATH,
+        ))
+    }
+
+    /// Try to report steady state.
+    pub(crate) fn report_steady(&self) -> Pin<Box<dyn Future<Output = Result<bool, Error>>>> {
+        trace!("filesystem strategy, report steady: {}", true);
+
+        let res = future::ok(true);
+        Box::pin(res)
+    }
+
+    /// Asynchronous helper function that returns a future indicating
+    /// whether finalization is allowed.
+    async fn filesystem_allow_finalization(finalization_marker_path: &str) -> Result<bool> {
+        if !verify_file_metadata(finalization_marker_path).await? {
+            return Ok(false);
+        }
+
+        if is_expired(finalization_marker_path).await? {
+            return Ok(false);
+        }
+
+        Ok(true)
+    }
+}
+
+/// Verify that finalization marker file exists, is a regular file,
+/// has the correct permissions, and is owned by the root user.
+#[context("failed to verify finalization marker file metadata")]
+async fn verify_file_metadata(path: &str) -> Result<bool> {
+    let attr = tokio_fs::metadata(path).await;
+    // If `path` doesn't exist, return false early.
+    if attr.is_err() {
+        return Ok(false);
+    }
+    let attr = attr.unwrap();
+    let mode = attr.permissions().mode();
+
+    // Octal 100644 represents regular file, rw-r--r--.
+    if mode != 0o100644 {
+        anyhow::bail!("incorrect file permissions");
+    }
+
+    if attr.uid() != 0 {
+        anyhow::bail!("file not owned by root user");
+    }
+
+    Ok(true)
+}
+
+/// Check whether the finalization marker file has expired, if `notAfter` key
+/// exists.
+async fn is_expired(path: &str) -> Result<bool> {
+    match parse_expiry_timestamp(path).await? {
+        Some(expiry_timestamp) => {
+            // We can `unwrap()` since we're certain `UNIX_EPOCH` is in the past.
+            let current_timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs();
+            if current_timestamp > expiry_timestamp {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+        None => Ok(false),
+    }
+}
+
+#[context("failed to parse expiry timestamp from marker file")]
+async fn parse_expiry_timestamp(path: &str) -> Result<Option<u64>> {
+    let data = tokio_fs::read_to_string(path).await?;
+    let marker_json: FinalizationMarker =
+        serde_json::from_str(&data).context("failed to parse JSON content")?;
```

Comment:
Again this doesn't matter here but with nontrivial data it's better to deserialize streaming; see https://github.com/serde-rs/json/issues/316 which can be worked easily enough in most cases (like this one) deserializing in a `spawn_blocking(||)`.

Reply:
I've just put a block of synchronous code in a block and put it inside a `spawn_blocking()`. Am I allowed to do this? 😅